_source/_posts/2023-01-23-keycloak-kubernetes-prod.md

```diff
@@ -0,0 +1,637 @@
+
+At a high level, the following steps are required for obtaining a certificate from an ACME server. In the test deployment, cert-manager acts as the agent, and Let's Encrypt staging as the CA (Certificate Authority).
+
+1. The agent registers to the CA with a key pair
+2. The agent submits an order for a certificate to be issued
+3. The CA issues a DNS01 or HTTP01 challenge
+4. The CA provides a nonce the agent must sign with its private key
+5. The agent completes the challenge
+6. The agent signs the nonce with its private key
+7. The agent notifies the CA it is ready for validation
+8. The CA verifies the signature on the nonce satisfied
+9. The CA verifies the challenge has been satisfied
+10. The authorized agent sends a certificate signing request
+11. The CA issues the certificate
+
+### About ACME orders and challenges
+
+When a `Certificate` resource is created, cert-manager automatically creates a `CertificateRequest`. An `Order` resource represents a single certificate request, also created automatically, and is used by the ACME issuer to manage the lifecycle of the ACME order for a signed TLS certificate. The `Order` resource manages challenges for the order with the `Challenge` resource.
+
+The `Challenge` resource is queued for scheduled processing. The challenge controller will perform a self-check, and once it is passed, the ACME authorization is accepted for the ACME server to perform the validation.
+
+### The HTTP01 Ingress solver
+
+The HTTP-01 challenge is the most common challenge type. Let's Encrypt gives a token to the agent, and the agent must place a file at http://<YOUR_DOMAIN>/.well-known/acme-challenge/<TOKEN> that contains the token and a thumbprint (computation) of the agent. Once the agent tells Let's Encrypt that the file is ready, Let's Encrypt validates the file and issues the certificate.
+
+The challenge type is configured in the issuer, along with the ingress class for the solver. cert-manager will create a new Ingress resource to route Let's Encrypt challenge requests to the solver pods, which are also created automatically.
+
+The following annotations are added by the k8s sub-generator in the ingress resource, for the challenge solver to work and the TLS certificate to be automatically created by cert-manager:
+
+- `kubernetes.io/ingress.allow-http: "true"`: Required to allow HTTP connections from challenge requests
+- `cert-manager.io/issuer: letsencrypt-staging`: The name of the issuer to acquire the certificate, which must be in the same namespace
+- `acme.cert-manager.io/http01-edit-in-place: "true"`: The ingress is modified in place, instead of creating a new ingress resource for the HTTP01 challenge
+- `cert-manager.io/issue-temporary-certificate: "true"`: A temporary certificate will be set on the secret until the final certificate has been returned. used for keeping compatibility with the `ingress-gce` component.
```

Comment:
Are all these settings in the same YAML file? If so, it might be easier to read by putting that code block here, with the explanations as comments. If they're not in the same file, please disregard.

Reply:
Updated